storage/src/binary/mod.rs

```diff
@@ -0,0 +1,457 @@
+//!
+//! This data structure is often used to generate a root for a block digest (over included transactions).
+//!
+//! # Example
+//!
+//! ```rust
+//! use commonware_storage::binary::Tree;
+//! use commonware_cryptography::{hash, Sha256, sha256::Digest};
+//!
+//! // Build tree
+//! let txs = [b"tx1", b"tx2", b"tx3", b"tx4"];
+//! let digests: Vec<Digest> = txs.iter().map(|tx| hash(*tx)).collect();
+//! let mut hasher = Sha256::default();
+//! let tree = Tree::new(&mut hasher, digests.clone()).unwrap();
+//! let root = tree.root();
+//!
+//! // Generate a proof for leaf at index 1.
+//! let proof = tree.prove(0).unwrap();
+//! assert!(proof.verify(&mut hasher, &digests[0], 0, &root));
+//! ```
+
+use commonware_cryptography::{Digest, Hasher};
+use std::mem::size_of;
+
+/// Combines two digests into a new digest.
+fn combine<H: Hasher>(hasher: &mut H, left: &H::Digest, right: &H::Digest) -> H::Digest {
+    hasher.update(left);
+    hasher.update(right);
+    hasher.finalize()
+}
+
+/// A stateless Binary Merkle Tree that computes a root over an arbitrary set
+/// of [commonware_cryptography::Digest].
+#[derive(Clone, Debug)]
+pub struct Tree<H: Hasher> {
+    /// Levels of the tree: level 0 contains the leaves, and each subsequent level
+    /// contains the parent nodes computed from the previous level.
+    ///
+    /// The last level contains a single node: the root.
+    levels: Vec<Vec<H::Digest>>,
+}
+
+impl<H: Hasher> Tree<H> {
+    /// Builds a Merkle Tree from a slice of leaf digests.
+    ///
+    /// If `leaves` is empty, returns `None`.
+    ///
+    /// # Warning
+    ///
+    /// It is not safe to insert `H::Digest::default()` as a leaf. This could lead to
+    /// a proof being generated for a non-existent leaf.
+    pub fn new(hasher: &mut H, leaves: Vec<H::Digest>) -> Option<Self> {
+        // Ensure there are non-zero leaves.
+        if leaves.is_empty() {
+            return None;
+        }
+
+        // Level 0: the leaves.
+        let mut levels = Vec::new();
+        levels.push(leaves);
+
+        // Build higher levels until we reach the root.
+        let default = H::Digest::default();
+        while levels.last().unwrap().len() > 1 {
+            let current_level = levels.last().unwrap();
+            let mut next_level = Vec::with_capacity((current_level.len() + 1) / 2);
+            for chunk in current_level.chunks(2) {
+                let left = &chunk[0];
+                let right = if chunk.len() == 2 {
+                    &chunk[1]
+                } else {
+                    // Use the default digest for the right child if no right child exists.
+                    &default
+                };
+                next_level.push(combine(hasher, left, right));
+            }
+            levels.push(next_level);
+        }
+        Some(Self { levels })
+    }
+
+    /// Returns a reference to the Merkle root, if the tree is non-empty.
```

Comment:
You say "returns a reference" but it's returning a clone?

Reply:
Yeah, this was a stale comment 👍 . Thanks!